File: binding/wire/wire.go
// Package wire implements the binary protocol spoken between the Go worker and
// the SQLite DB worker. See docs/PROTOCOL.md, which this package is the
// normative Go half of.
//
// It carries no build tag on purpose: the codec is plain byte manipulation, so
// `go test ./...` exercises it on the host without a browser.
package wire

// Version is the protocol version carried in every frame header. Any change to
// the opcodes, payloads or flags in docs/PROTOCOL.md must bump it.
const Version = 1

// HeaderSize is the fixed size of a frame header in bytes.
const HeaderSize = 8

// Op identifies a frame's kind. Requests are < 0x80, responses >= 0x80.
type Op uint8

const (
	OpOpen      Op = 0x01
	OpClose     Op = 0x02
	OpPrepare   Op = 0x03
	OpFinalize  Op = 0x04
	OpQuery     Op = 0x05
	OpQueryStmt Op = 0x06
	OpExec      Op = 0x07
	OpExecStmt  Op = 0x08
	OpCredit    Op = 0x09
	OpAbort     Op = 0x0A
	OpShutdown  Op = 0x0B

	OpReady      Op = 0x80
	OpOK         Op = 0x81
	OpError      Op = 0x82
	OpOpened     Op = 0x83
	OpPrepared   Op = 0x84
	OpRows       Op = 0x85
	OpExecResult Op = 0x86
	OpAborted    Op = 0x87
)

func (o Op) String() string {
	switch o {
	case OpOpen:
		return "OPEN"
	case OpClose:
		return "CLOSE"
	case OpPrepare:
		return "PREPARE"
	case OpFinalize:
		return "FINALIZE"
	case OpQuery:
		return "QUERY"
	case OpQueryStmt:
		return "QUERY_STMT"
	case OpExec:
		return "EXEC"
	case OpExecStmt:
		return "EXEC_STMT"
	case OpCredit:
		return "CREDIT"
	case OpAbort:
		return "ABORT"
	case OpShutdown:
		return "SHUTDOWN"
	case OpReady:
		return "READY"
	case OpOK:
		return "OK"
	case OpError:
		return "ERROR"
	case OpOpened:
		return "OPENED"
	case OpPrepared:
		return "PREPARED"
	case OpRows:
		return "ROWS"
	case OpExecResult:
		return "EXEC_RESULT"
	case OpAborted:
		return "ABORTED"
	}
	return "Op(" + itoa(int(o)) + ")"
}

// Frame flags.
const (
	// FlagEOF marks the last frame for a request id.
	FlagEOF uint16 = 1 << 0
	// FlagHasColumns marks a ROWS frame that is prefixed with a columns block.
	FlagHasColumns uint16 = 1 << 1
)

// Tag identifies the storage class of an encoded value. It mirrors SQLite's
// storage classes rather than its SQLITE_* constants, so that the wire format
// does not depend on them.
type Tag uint8

const (
	TagNull Tag = 0x00
	TagInt  Tag = 0x01
	TagReal Tag = 0x02
	TagText Tag = 0x03
	TagBlob Tag = 0x04
)

func (t Tag) String() string {
	switch t {
	case TagNull:
		return "NULL"
	case TagInt:
		return "INT"
	case TagReal:
		return "REAL"
	case TagText:
		return "TEXT"
	case TagBlob:
		return "BLOB"
	}
	return "Tag(" + itoa(int(t)) + ")"
}

// absentLen marks a nullable string as absent. It is distinct from length 0,
// which is a present empty string.
const absentLen = 0xFFFFFFFF

// Capability bits reported in the READY frame.
const (
	CapCrossOriginIsolated uint32 = 1 << 0
	CapSharedArrayBuffer   uint32 = 1 << 1
	CapBigInt              uint32 = 1 << 2
	CapProgressHandler     uint32 = 1 << 3
	CapVFSOpfs             uint32 = 1 << 4
	CapVFSOpfsSAHPool      uint32 = 1 << 5
	CapVFSMemdb            uint32 = 1 << 6
)

// Header is the fixed 8-byte prefix of every frame.
type Header struct {
	Version uint8
	Op      Op
	Flags   uint16
	ID      uint32
}

func (h Header) EOF() bool        { return h.Flags&FlagEOF != 0 }
func (h Header) HasColumns() bool { return h.Flags&FlagHasColumns != 0 }

func itoa(n int) string {
	if n == 0 {
		return "0"
	}
	neg := n < 0
	if neg {
		n = -n
	}
	var b [20]byte
	i := len(b)
	for n > 0 {
		i--
		b[i] = byte('0' + n%10)
		n /= 10
	}
	if neg {
		i--
		b[i] = '-'
	}
	return string(b[i:])
}

File: src/wire.ts
/**
 * The binary protocol spoken between the Go worker and the DB worker.
 *
 * This is the TypeScript half of docs/PROTOCOL.md; `binding/wire` is the Go
 * half. Both are checked against `testdata/wire/vectors.json`, which is
 * produced by a third, independent implementation
 * (`scripts/gen-wire-vectors.py`) so that a bug in either half cannot be
 * blessed as the specification.
 *
 * The encoder sits in the row loop, so it avoids per-value allocation: one
 * growable buffer per frame, written through a single DataView.
 */

export const PROTOCOL_VERSION = 1
export const HEADER_SIZE = 8

export const Op = {
	OPEN: 0x01,
	CLOSE: 0x02,
	PREPARE: 0x03,
	FINALIZE: 0x04,
	QUERY: 0x05,
	QUERY_STMT: 0x06,
	EXEC: 0x07,
	EXEC_STMT: 0x08,
	CREDIT: 0x09,
	ABORT: 0x0a,
	SHUTDOWN: 0x0b,

	READY: 0x80,
	OK: 0x81,
	ERROR: 0x82,
	OPENED: 0x83,
	PREPARED: 0x84,
	ROWS: 0x85,
	EXEC_RESULT: 0x86,
	ABORTED: 0x87,
} as const

export type Op = (typeof Op)[keyof typeof Op]

export const OP_NAME: Record<number, string> = Object.fromEntries(
	Object.entries(Op).map(([k, v]) => [v, k]),
)

export const Flag = {
	EOF: 1 << 0,
	HAS_COLUMNS: 1 << 1,
} as const

/** Storage classes, mirroring SQLite's without depending on its constants. */
export const Tag = {
	NULL: 0x00,
	INT: 0x01,
	REAL: 0x02,
	TEXT: 0x03,
	BLOB: 0x04,
} as const

export type Tag = (typeof Tag)[keyof typeof Tag]

/** Marks a nullable string as absent, which is distinct from length 0. */
const ABSENT = 0xffffffff

export const Cap = {
	CROSS_ORIGIN_ISOLATED: 1 << 0,
	SHARED_ARRAY_BUFFER: 1 << 1,
	BIGINT: 1 << 2,
	PROGRESS_HANDLER: 1 << 3,
	VFS_OPFS: 1 << 4,
	VFS_OPFS_SAHPOOL: 1 << 5,
	VFS_MEMDB: 1 << 6,
} as const

export type Header = {
	version: number
	op: number
	flags: number
	id: number
}

export class WireError extends Error {
	constructor(message: string) {
		super(message)
		this.name = 'WireError'
	}
}

const encoder = new TextEncoder()
const decoder = new TextDecoder()

export class FrameWriter {
	private buf: Uint8Array
	private view: DataView
	private n = HEADER_SIZE

	constructor(op: number, flags = 0, id = 0, hint = 64) {
		this.buf = new Uint8Array(HEADER_SIZE + hint)
		this.view = new DataView(this.buf.buffer)
		this.view.setUint8(0, PROTOCOL_VERSION)
		this.view.setUint8(1, op)
		this.view.setUint16(2, flags, true)
		this.view.setUint32(4, id, true)
	}

	/** Current frame size, header included. */
	get length(): number {
		return this.n
	}

	/**
	 * Batch encoders do not know whether a frame is the last one until they
	 * have filled it, so flags stay rewritable.
	 */
	setFlags(flags: number): void {
		this.view.setUint16(2, flags, true)
	}

	/**
	 * The encoded frame, as a view onto the (possibly larger) backing buffer.
	 * Post the view and transfer `view.buffer`: Go's js.CopyBytesToGo accepts
	 * only Uint8Array, and it honours byteOffset, so no copy is needed here.
	 * Transferring detaches the buffer, so the writer must not be reused.
	 */
	frame(): Uint8Array {
		return this.buf.subarray(0, this.n)
	}

	private grow(need: number): void {
		if (this.n + need <= this.buf.length) return
		let cap = this.buf.length * 2
		while (cap < this.n + need) cap *= 2
		const next = new Uint8Array(cap)
		next.set(this.buf.subarray(0, this.n))
		this.buf = next
		this.view = new DataView(next.buffer)
	}

	/**
	 * Reserves a u32 and returns its offset. The row encoder cannot know a
	 * batch's row count until the batch is full, but the count is specified to
	 * come before the rows, so it is patched in at flush time.
	 */
	reserveU32(): number {
		const at = this.n
		this.u32(0)
		return at
	}

	patchU32(at: number, v: number): void {
		this.view.setUint32(at, v, true)
	}

	u8(v: number): void {
		this.grow(1)
		this.view.setUint8(this.n, v)
		this.n += 1
	}

	bool(v: boolean): void {
		this.u8(v ? 1 : 0)
	}

	u16(v: number): void {
		this.grow(2)
		this.view.setUint16(this.n, v, true)
		this.n += 2
	}

	u32(v: number): void {
		this.grow(4)
		this.view.setUint32(this.n, v, true)
		this.n += 4
	}

	i32(v: number): void {
		this.grow(4)
		this.view.setInt32(this.n, v, true)
		this.n += 4
	}

	/**
	 * Accepts a bigint so `sqlite3_column_int64` can be written straight
	 * through: a BigInt must never reach Go, where syscall/js panics on it.
	 */
	i64(v: bigint | number): void {
		this.grow(8)
		this.view.setBigInt64(this.n, typeof v === 'bigint' ? v : BigInt(v), true)
		this.n += 8
	}

	f64(v: number): void {
		this.grow(8)
		this.view.setFloat64(this.n, v, true)
		this.n += 8
	}

	/** Length-prefixed raw bytes. */
	bytes(v: Uint8Array): void {
		this.u32(v.length)
		this.grow(v.length)
		this.buf.set(v, this.n)
		this.n += v.length
	}

	/** Length-prefixed UTF-8 string. */
	str(v: string): void {
		// A UTF-8 encoding is at most 3 bytes per UTF-16 code unit, and
		// surrogate pairs cost 4 bytes for 2 units, so 3x is a safe bound.
		this.grow(4 + v.length * 3)
		const at = this.n
		this.n += 4
		const { written } = encoder.encodeInto(v, this.buf.subarray(this.n))
		this.view.setUint32(at, written, true)
		this.n += written
	}

	/** Nullable string; absent is distinct from empty. */
	nstr(v: string | null | undefined): void {
		if (v == null) {
			this.u32(ABSENT)
			return
		}
		this.str(v)
	}

	valueNull(): void {
		this.u8(Tag.NULL)
	}

	valueInt(v: bigint | number): void {
		this.u8(Tag.INT)
		this.i64(v)
	}

	valueReal(v: number): void {
		this.u8(Tag.REAL)
		this.f64(v)
	}

	valueText(v: string): void {
		this.u8(Tag.TEXT)
		this.str(v)
	}

	/**
	 * TEXT whose bytes came straight off the wasm heap. SQLite does not
	 * guarantee TEXT is valid UTF-8, and reading it as bytes also avoids a
	 * UTF-8 -> UTF-16 -> UTF-8 round trip and truncation at an embedded NUL.
	 */
	valueTextBytes(v: Uint8Array): void {
		this.u8(Tag.TEXT)
		this.bytes(v)
	}

	valueBlob(v: Uint8Array): void {
		this.u8(Tag.BLOB)
		this.bytes(v)
	}
}

export class FrameReader {
	private view: DataView
	private i = 0

	private buf: Uint8Array

	constructor(buf: Uint8Array, start = 0) {
		this.buf = buf
		this.view = new DataView(buf.buffer, buf.byteOffset, buf.byteLength)
		this.i = start
	}

	/** Decodes the header and returns a reader positioned at the payload. */
	static header(buf: Uint8Array): { header: Header; reader: FrameReader } {
		if (buf.length < HEADER_SIZE) throw new WireError('frame truncated')
		const view = new DataView(buf.buffer, buf.byteOffset, buf.byteLength)
		const header: Header = {
			version: view.getUint8(0),
			op: view.getUint8(1),
			flags: view.getUint16(2, true),
			id: view.getUint32(4, true),
		}
		if (header.version !== PROTOCOL_VERSION) {
			throw new WireError(
				`unsupported frame version ${header.version}, want ${PROTOCOL_VERSION}`,
			)
		}
		return { header, reader: new FrameReader(buf, HEADER_SIZE) }
	}

	get remaining(): number {
		return this.buf.length - this.i
	}

	private need(n: number): number {
		if (this.remaining < n) throw new WireError('frame truncated')
		const at = this.i
		this.i += n
		return at
	}

	u8(): number {
		return this.view.getUint8(this.need(1))
	}

	bool(): boolean {
		return this.u8() !== 0
	}

	u16(): number {
		return this.view.getUint16(this.need(2), true)
	}

	u32(): number {
		return this.view.getUint32(this.need(4), true)
	}

	i32(): number {
		return this.view.getInt32(this.need(4), true)
	}

	i64(): bigint {
		return this.view.getBigInt64(this.need(8), true)
	}

	f64(): number {
		return this.view.getFloat64(this.need(8), true)
	}

	/** Length-prefixed bytes, as a subarray of the frame — not a copy. */
	bytes(): Uint8Array {
		const n = this.u32()
		if (n === ABSENT) {
			throw new WireError('absent marker in a non-nullable field')
		}
		const at = this.need(n)
		return this.buf.subarray(at, at + n)
	}

	str(): string {
		return decoder.decode(this.bytes())
	}

	nstr(): string | null {
		const n = this.u32()
		if (n === ABSENT) return null
		const at = this.need(n)
		return decoder.decode(this.buf.subarray(at, at + n))
	}

	tag(): Tag {
		const t = this.u8()
		if (t > Tag.BLOB) {
			throw new WireError(`unknown value tag 0x${t.toString(16).padStart(2, '0')}`)
		}
		return t as Tag
	}
}
